File: JsonGson/app/src/main/java/com/example/jsongson/MainActivity.java
package com.example.jsongson;

import androidx.appcompat.app.AppCompatActivity;

import android.content.ClipData;
import android.os.Bundle;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

public class MainActivity extends AppCompatActivity {

    ItemData itemData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        itemData = new ItemData("mike", 10);

        Gson gson = new Gson();

        String alpha = gson.toJson(itemData); // toJson은 Java-Json의 변환인데 Json으로 변환을 한다는게 String Type으로 바꾸는 것 같다

        Log.i("TAG", alpha); // Json 변환시 {"num":0}으로 변환됨.

        ItemData revItemData = gson.fromJson(alpha, ItemData.class);

        Log.i("TAG", revItemData.setData());

        if(false){
            Log.i("TAG","false");
        }else{
            Log.i("TAG","true");

        }
    }
}

File: Retrofit21/app/src/main/java/com/puresoftware/retrofit2_1/MainActivity.kt
package com.puresoftware.retrofit2_1

// https://jaejong.tistory.com/33
// https://jaejong.tistory.com/38?category=873924
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import retrofit2.*
import retrofit2.converter.gson.GsonConverterFactory

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // instance
        var retrofit: Retrofit = Retrofit.Builder().baseUrl("https://jsonplaceholder.typicode.com")
            .addConverterFactory(GsonConverterFactory.create()).build()
        var retrofitService: RetrofitService = retrofit.create(RetrofitService::class.java)

        // 일반적인 get
        var call = retrofitService.getPosts()
        call.enqueue(object :Callback<List<PostResult>>{

            override fun onResponse(
                call: Call<List<PostResult>>, response: Response<List<PostResult>>) {
                Log.i("TAG","성공")
            }

            override fun onFailure(call: Call<List<PostResult>>, t: Throwable) {
                Log.i("TAG","실패")
            }
        })

        // path get
        var call2 = retrofitService.getPost("5")
        call2.enqueue(object :Callback<List<PostResult>>{

            override fun onResponse(
                call: Call<List<PostResult>>, response: Response<List<PostResult>>) {
                Log.i("TAG","성공")
            }
            override fun onFailure(call: Call<List<PostResult>>, t: Throwable) {
                Log.i("TAG","실패")
            }

        })

        // Query get
        var call3 = retrofitService.getPostsQuery("10")
        call3.enqueue(object :Callback<List<PostResult>>{

            override fun onResponse(
                call: Call<List<PostResult>>, response: Response<List<PostResult>>) {
                Log.i("TAG","성공")
            }

            override fun onFailure(call: Call<List<PostResult>>, t: Throwable) {
                Log.i("TAG","실패")
            }
        })

        // MultiQuery
        var call4 = retrofitService.getPostsMultiQuery("10","96")
        call4.enqueue(object:Callback<List<PostResult>>{

            override fun onResponse(
                call: Call<List<PostResult>>, response: Response<List<PostResult>>) {
                Log.i("TAG","성공")
            }

            override fun onFailure(call: Call<List<PostResult>>, t: Throwable) {
                Log.i("TAG","실패")
            }
        })

        // post
        val post = PostResult(1000,"test title","test body")
        var call5 = retrofitService.setPostBody(post)

//        var call6
        // git test
        // git test!!!!
        // git commit test!


    }
}